---
title: "Make figures for the paper"
author: "Siyao Liu, Ph.D. | Lineberger Bioinformatics Core"
date: "`r format(Sys.time(), '%d %B %Y')`"
output: 
  rmdformats::downcute:
      self_contained: true
      thumbnails: false
      lightbox: true
      gallery: false
      highlight: kate
      toc_float: yes
      toc_depth: 3
      theme: cosmo
      code_folding: hide
params:
  r33_data: "/datastore/lbcfs/labs/bioinformatics/wdgong/projects/Nancy_Thomas/TIL-prediction/data/r33_preProcessed.Rdata"
  til_image: "/datastore/lbcfs/labs/bioinformatics/wdgong/projects/Nancy_Thomas/TIL-prediction/data/Alg5_selectedROI_091719.txt" # Image data from Shannon #ProcessedFiles_071118/CellrAnalysis_per_Slide_20170103-Alg4_350u_PGEM1599excluded_273added_groupByMaxNumberNucleus.txt
  estF: "/datastore/lbcfs/labs/bioinformatics/wdgong/projects/Nancy_Thomas/TIL-prediction/results/Deconvolution/MethylDecon.rds" # data matrix from ../results/Deconvolution/Deconvolution_PGEM.Rmd
# Use EpiDISH (RPC, CBS and CP) and MethylResolver to estimate T cell percentage
# EpiDISH_CP provides the best R2 with IHC measured CD3+ T cell percentage (See ../results/Deconvolution/Decon.Til.scatter.pdf)
  BestAlpha: 0.1
  process_dir: "./processedData/"
  results_dir: "./results/"
  colors_file: "./scripts/R/colors.R"
  
  final_probes: "/datastore/nextgenout5/share/labs/bioinformatics/siyao/projects/Thomas-Edmiston_20240711_PredictTILs/results/383k_FinalModel_logY_c0.01_alpha0.1/glmnet_log_y_c0.01_FinalModel_alpha0.1_ProbeCoef.txt"
  
  tcga_dir: "/datastore/lbcfs/labs/bioinformatics/wdgong/projects/Nancy_Thomas/TIL-prediction/results/TCGA_SKCM/"
  
  intermel.data: "/datastore/nextgenout5/share/labs/bioinformatics/vondras/melanoma/share_data/share_10_17_2022/IML10172022.RData"
  intermel.clin: "/datastore/nextgenout5/share/labs/bioinformatics/vondras/melanoma/data/IML_clinical_metadata.RData"

  knit_rdir: "/datastore/nextgenout5/share/labs/bioinformatics/siyao/projects/Thomas-Edmiston_20240711_PredictTILs/" 
knit: (function(input_file, encoding) {
  rmarkdown::render(input_file, 
                              output_dir="/datastore/nextgenout5/share/labs/bioinformatics/siyao/projects/Thomas-Edmiston_20240711_PredictTILs/")}) 
editor_options: 
  markdown: 
    wrap: 72
---

```{=html}
<style>
mark {
  background-color: #E6E6FA;
  color: #000;
  border: 1px solid #E6E6FA;
  border-radius: 5px;
  font-family: monospace;
  
}

.page-content code {
  background: #E6E6FA;
  color: #000;
}

code {
  background: #E6E6FA;
  background-color: #E6E6FA;
  color: #000;
  border: 1px solid #E6E6FA;
  border-radius: 5px;
  font-family: monospace;
}

code-out {
  background-color: #E6E6FA;
  color: #000;
  border: 1px solid #E6E6FA;
  border-radius: 2px;
}
</style>
```


# Purpose

The objectives of this analysis are: <break>

Make Figures for the manuscript.

# Set up

```{r, setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, eval = TRUE, message = FALSE, warning = FALSE)
# root dir is for input and necessary files 
# knitr outdir (html) is set in yaml header
# setting root.dir here creates process.dir and results.dir in user directory
# but will cause issues with input files if absolute paths are not used.
knitr::opts_knit$set(root.dir = params$knit_rdir)
```

## Load libraries

```{r, load-libraries}
library(stringr)
library(kableExtra)
library(caret)
library(glmnet)
library(dplyr)
library(ggplot2)
library(RColorBrewer)
library(ComplexHeatmap)
library(circlize)
library(ggrepel)
library(reshape2)
library(randomForest)
library(xgboost)
library(cowplot)
library(grid)
library(gridExtra)
library(data.table)
library(survminer)
library(survival)
library(EpiDISH)
library(ggpubr)
library(patchwork)
```

## Functions

```{r, load-functions}
source("./scripts/R/Functions.R")
```

## Colors

```{r, define-colors}
til.col <- RColorBrewer::brewer.pal(3, "Set1")
names(til.col) <- c("absent", "nonbrisk", "brisk")
```


## Parameters

```{r, echo=FALSE, results="asis"}
p <- unlist(list(params), recursive = FALSE)
params.df <- data.frame(Parameter=names(p), Value=as.character(p))
kable(params.df, booktabs = TRUE) %>%
  kableExtra::kable_styling(font_size = 12)
```



# Boxplot - Ground truth vs. Epi-TIL grouped by TIL grade using final Epi-TIL model

```{r, fig.width=6, fig.height=5}
# load r33 preprocessed data
load(params$r33_data)
y$r33_Sample_ID <- paste("X", y$r33_Sample_ID,".AVG_Beta", sep="")
y <- y[y$r33_Sample_ID %in% colnames(xb), ]
xb <- xb[, y$r33_Sample_ID]

sub.xb <- readRDS(file="/datastore/nextgenout5/share/labs/bioinformatics/siyao/projects/Siyao_20230908_PredictTILs/processedData/231002_r33_383kProbes_Til-perc.rds")
all.x <- sub.xb[, -which(colnames(sub.xb) %in% c("outcome", "outcome_cat", "TotalT"))]

# response: CD3TIL perc
all.y <- sub.xb$outcome
# log10 transform y since y looks right skewed
C <- 0.01
all.y <- log10(all.y + C)
```

## Calculate Epi-TIL score

```{r}
model.probe <- read.table("/datastore/nextgenout5/share/labs/bioinformatics/siyao/projects/Thomas-Edmiston_20240711_PredictTILs/results/383k_FinalModel_logY_c0.01_alpha0.1/glmnet_log_y_c0.01_FinalModel_alpha0.1_ProbeCoef.txt", sep = "\t", header=TRUE)
dim(model.probe)
model.probe$probe <- rownames(model.probe)

# Subset data
tmpd <- t(all.x)[ rownames(t(all.x)) %in% rownames(model.probe), ]
#dim(tmpd)
##[1] 128  80

# Subset weights and convert tibble to named vector
tmpp <- model.probe[model.probe$probe %in% rownames(tmpd), ]
#dim(tmpp)
tmpp2 <- as.numeric(tmpp$s1)
names(tmpp2) <- tmpp$probe    

# Check if the subsetted probes are in the same order
#identical(names(tmpp2), rownames(tmpd))
## TRUE

# Reorder
tmpp2 <- tmpp2[match(rownames(tmpd), names(tmpp2))]
#identical(names(tmpp2), rownames(tmpd))
##TRUE

# Prediction calculation
datp <- t(tmpd) %*% tmpp2
datp <- datp + as.numeric(model.probe[model.probe$probe == "(Intercept)", "s1"])
#dim(datp)

# Reformat datp
datp_table <- data.frame("Sample_Name"=rownames(datp), 
                         "glmnet.pred" = as.numeric(datp))
row.names(datp_table) <- NULL
datp_table$glmnet.pred.tertile <- dplyr::ntile(datp_table$glmnet.pred, 3)
table(datp_table$glmnet.pred.tertile)
```


## Plot Epi-TIL vs. ground truth in each TIL grade category

```{r, fig.width=8, fig.height=5}
# get final predictions
identical(datp_table$Sample_Name, rownames(all.x))
pred.df <- data.frame("Name"=datp_table$Sample_Name, 
                      "TIL"=sub.xb$outcome,
                      "log_TIL"=all.y,
                      "log_Pred"=datp_table$glmnet.pred,
                       "Pred"=10^(datp_table$glmnet.pred) - 0.01)
# merge pred.df with clinical data
merge.df <- merge(pred.df, 
                  y, 
                  by.x="Name", by.y="r33_Sample_ID", sort=FALSE)
table(merge.df$tilcat2)

merge.df$tilcat3 <- plyr::mapvalues(
    merge.df$tilcat2,
    from = c(0, 1, 2),
    to = c("absent", "nonbrisk", "brisk")
  )
table(merge.df$tilcat3)
sum(is.na(merge.df$tilcat3))

# remove NA in til grade
merge.df <- merge.df[!is.na(merge.df$tilcat3),]
dim(merge.df)

plot.df <- merge.df[, c("tilcat3", "TIL", "Pred", "log_TIL", "log_Pred")]
plot.df.long <- melt(plot.df, id.vars = "tilcat3")
plot.df.long$tilcat3 <- factor(plot.df.long$tilcat3, levels = c("absent", "nonbrisk", "brisk"))
plot.df.long$variable <- factor(plot.df.long$variable, levels=c("TIL", "Pred", "log_TIL", "log_Pred"))

# eliminate one indeterminate sample
p1 <- ggplot(plot.df.long[!is.na(plot.df.long$tilcat3)&plot.df.long$variable%in%c("TIL","Pred"), ], 
             aes(x=tilcat3, y=value, fill=variable)) +
  geom_boxplot(outlier.shape = NA, position = position_dodge()) +
  geom_point(size = 1, color="grey35", position = position_jitterdodge()) +
  ggtitle(paste0("Epi-TIL Final model: ", (dim(model.probe)[1]-1), " probes")) +
  scale_fill_manual("", values=c("tomato2", "steelblue2"), 
                    labels=c("Epi-TIL", "Ground truth")) +
  theme_45() + 
  xlab("") + 
  ylab("Proportion of CD3+ T cell")
p1

# eliminate one indeterminate sample
p2 <- ggplot(plot.df.long[!is.na(plot.df.long$tilcat3)&plot.df.long$variable%in%c("log_TIL","log_Pred"), ], 
             aes(x=tilcat3, y=value, fill=variable)) +
  geom_boxplot(outlier.shape = NA, position = position_dodge()) +
  geom_point(size = 1, color="grey35", position = position_jitterdodge()) +
  ggtitle(paste0("Epi-TIL Final model: ", (dim(model.probe)[1]-1), " probes")) +
  scale_fill_manual("", values=c("tomato2", "steelblue2"), 
                    labels=c("Epi-TIL", "Ground truth")) +
  theme_45() + 
  xlab("") + 
  ylab("log10 (Prop of CD3+ T cell + 0.01)")
p2
```



# PGEM: Epi-TIL predictors to DSS

## DSS KM by Epi-TIL tertiles

```{r, prep-KM}
identical(pred.df$Name, y$r33_Sample_ID)
##FALSE
sub.y <- y[match(pred.df$Name, y$r33_Sample_ID), ]
sub.y$first_recurr_date <- as.Date(as.character(sub.y$first_recurr_date), format = "%m/%d/%Y")
sub.y$Date_Dx <- as.Date(as.character(sub.y$Date_Dx), format = "%m/%d/%Y")
sub.y$current_status_date <- as.Date(as.character(sub.y$current_status_date), format = "%m/%d/%Y")
identical(sub.y$r33_Sample_ID, pred.df$Name)

predictions <- data.frame(
  Sample = rownames(sub.y),
  prog_event = (sub.y$first_recurrence != "" & sub.y$first_recurrence != "*see note"),
  dwd_event = sub.y$current_status == "DWD" | sub.y$current_status == "DOD",
  stage = sub.y$stage_t,
  age = sub.y$age_dx,
  sex = sub.y$sex,
  til_grade = sub.y$til,
  stringsAsFactors = F
)
predictions$stage <- factor(predictions$stage, levels = c("1a", "1b", "2a", "2b", "3a", "3b", "4a", "4b"), ordered = T)
predictions$prog_time <- as.numeric(sub.y$current_status_date - sub.y$Date_Dx)/30
predictions$prog_time[predictions$prog_event] <- as.numeric(sub.y$first_recurr_date[predictions$prog_event] - 
                                                              sub.y$Date_Dx[predictions$prog_event])/30
predictions$dwd_time <- as.numeric(sub.y$current_status_date - sub.y$Date_Dx)/30

identical(as.character(predictions$Sample), names(pred.df$Name))
##FALSE

predictions <- merge(predictions, pred.df, by.x="Sample", by.y="Name", sort=FALSE)
predictions$Pred.tertile <- dplyr::ntile(as.numeric(predictions$Pred), 3)
predictions$Pred.tertile2 <- dplyr::recode(predictions$Pred.tertile, 
                                       "1"="low", "2"="med", "3"="high" )
predictions$Pred.tertile2 <- factor(predictions$Pred.tertile2,
                                           levels=c("low","med","high"))

predictions$Pred.tertile.log <- dplyr::ntile(as.numeric(predictions$log_Pred), 3)
predictions$Pred.tertile2.log <- dplyr::recode(predictions$Pred.tertile.log, 
                                       "1"="low", "2"="med", "3"="high" )
predictions$Pred.tertile2.log <- factor(predictions$Pred.tertile2.log,
                                           levels=c("low","med","high"))


dim(predictions)

sum(is.na(predictions$til_grade))
##[1] 0
table(predictions$til_grade)
```

## Clean up 

```{r}
sum(is.na(predictions$dwd_time))
table(predictions$til_grade)

# remove sample with til grade = indeterminate
predictions <- predictions[predictions$til_grade!="indeterminate", ]
# remove dwd_time is NA
predictions <- predictions[!is.na(predictions$dwd_time), ]
dim(predictions)
##[1] 78 13
levels(predictions$sex)
levels(predictions$til_grade)
# drop levels
predictions$sex <- droplevels(predictions$sex)
predictions$til_grade <- droplevels(predictions$til_grade)
```


```{r}
tmp.logrank <- survival::survdiff(Surv(dwd_time, dwd_event) ~ Pred.tertile2, data=predictions)
tmp.p.logrank <- 1 - pchisq(tmp.logrank$chisq, length(tmp.logrank$n) - 1)
tmp.p.logrank
```


```{r, fig.width=7, fig.height=6.5}
pred.surv <- survival::survfit(Surv(dwd_time, dwd_event) ~ Pred.tertile2, data=predictions)
names(pred.surv$strata) <- gsub("Pred.tertile2=", "", names(pred.surv$strata))
dss.epi.til1 <- ggsurvplot(fit=pred.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #pval = signif(tmp.p.logrank,3), 
                                #palette = "Dark2",
                                palette = c("chartreuse4", "darkblue", "firebrick"),
                                ylab="Probability of DSS",
                                xlab="Time (months)", 
                                #xlim = c(0, 60), break.time.by = 12,
                                legend.title = "Epi-TIL predictions")
plot.new() 
print(dss.epi.til1, newpage = FALSE)
```


```{r, fig.width=7, fig.height=6.5}
pred.surv <- survival::survfit(Surv(dwd_time, dwd_event) ~ Pred.tertile2.log, data=predictions)
names(pred.surv$strata) <- gsub("Pred.tertile2.log=", "", names(pred.surv$strata))
dss.epi.til2 <- ggsurvplot(fit=pred.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #pval = signif(tmp.p.logrank,3), 
                                palette = c("chartreuse4", "darkblue", "firebrick"), 
                                #palette = "Dark2",
                                ylab="Probability of DSS",
                                xlab="Time (months)", 
                                #xlim = c(0, 60), break.time.by = 12,
                                legend.title = "Epi-TIL predictions (log10)")


plot.new() 
print(dss.epi.til2, newpage = FALSE)
```


## DSS KM by TIL grade

```{r, plotDSS-til, fig.width=7, fig.height=6.5}
predictions$til_grade <- factor(predictions$til_grade, 
                                levels=c("absent", "non-brisk", "brisk"))
til.surv <- survival::survfit(Surv(dwd_time, dwd_event) ~ til_grade, data=predictions)
names(til.surv$strata) <- gsub("til_grade=", "", names(til.surv$strata))

dss.til <- ggsurvplot(fit=til.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #pval = signif(tmp.p.logrank,3), 
                                palette = c("chartreuse4", "darkblue", "firebrick"), 
                                #palette = "Dark2",
                                ylab="Probability of DSS",
                                xlab="Time (months)", 
                                #xlim = c(0, 60), break.time.by = 12,
                                legend.title = "TIL grade")
dss.til
```




## DSS KM by EpiDISH_RPC tertiles

```{r, fig.width=7, fig.height=6.5}
m.decon <- readRDS(file="./results/RunMethylDeconOutput.rds")
# remove methylResolver
m.decon <- m.decon[-4]
# Calculate tertiles
m.decon$EpiDISH_RPC$TotalT.tertile <- ntile(m.decon$EpiDISH_RPC$TotalT, 3)
m.decon$EpiDISH_CBS$TotalT.tertile <- ntile(m.decon$EpiDISH_CBS$TotalT, 3)
m.decon$EpiDISH_CP$TotalT.tertile <- ntile(m.decon$EpiDISH_CP$TotalT, 3)
m.decon$EpiDISH_RPC$TotalT.tertile2 <- recode(m.decon$EpiDISH_RPC$TotalT.tertile,
                                              "1"="low", "2"="med", "3"="high")
m.decon$EpiDISH_RPC$TotalT.tertile2 <- factor(m.decon$EpiDISH_RPC$TotalT.tertile2,
                                              levels=c("low","med","high"))
m.decon$EpiDISH_CBS$TotalT.tertile2 <- recode(m.decon$EpiDISH_CBS$TotalT.tertile,
                                              "1"="low", "2"="med", "3"="high")
m.decon$EpiDISH_CBS$TotalT.tertile2 <- factor(m.decon$EpiDISH_CBS$TotalT.tertile2,
                                              levels=c("low","med","high"))
m.decon$EpiDISH_CP$TotalT.tertile2 <- recode(m.decon$EpiDISH_CP$TotalT.tertile,
                                              "1"="low", "2"="med", "3"="high")
m.decon$EpiDISH_CP$TotalT.tertile2 <- factor(m.decon$EpiDISH_CP$TotalT.tertile2,
                                              levels=c("low","med","high"))

all(predictions$Sample==rownames(m.decon[[1]]))
##FALSE

# remove the 2 samples filtered out earlier
m.decon[[1]] <- m.decon[[1]][rownames(m.decon[[1]])%in% predictions$Sample, ]
all(predictions$Sample==rownames(m.decon[[1]]))
m.decon[[2]] <- m.decon[[2]][rownames(m.decon[[2]])%in% predictions$Sample, ]
all(predictions$Sample==rownames(m.decon[[2]]))
m.decon[[3]] <- m.decon[[3]][rownames(m.decon[[3]])%in% predictions$Sample, ]
all(predictions$Sample==rownames(m.decon[[3]]))


# add EpiDISH results to "predictions"
predictions$EpiDISH.RPC.TotalT <- m.decon[[1]]$TotalT
predictions$EpiDISH.CBS.TotalT <- m.decon[[2]]$TotalT
predictions$EpiDISH.CP.TotalT <- m.decon[[3]]$TotalT

predictions$EpiDISH.RPC.TotalT.tertile <- m.decon[[1]]$TotalT.tertile2
predictions$EpiDISH.CBS.TotalT.tertile <- m.decon[[2]]$TotalT.tertile2
predictions$EpiDISH.CP.TotalT.tertile <- m.decon[[3]]$TotalT.tertile2

epidish.surv <- survival::survfit(Surv(dwd_time, dwd_event) ~ EpiDISH.RPC.TotalT.tertile, data=predictions)
names(epidish.surv$strata) <- gsub("EpiDISH.RPC.TotalT.tertile=", "", names(epidish.surv$strata))
dss.rpc <- survminer::ggsurvplot(fit = epidish.surv, 
                                 data = predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #pval = signif(tmp.p.logrank,3), 
                                palette = c("chartreuse4", "darkblue", "firebrick"), 
                                #palette = "Dark2",
                                ylab="Probability of DSS",
                                xlab="Time (months)", 
                                #xlim = c(0, 60), break.time.by = 12,
                                legend.title = "EpiDISH.RPC.TotalT")
dss.rpc
```


## Save KM pots

```{r}
pdf("./results/Figures/250203_PGEM_DSS_EpiTIL_KMplots_78_samples.pdf", width=7, height=6.5)
plot.new() 
print(dss.epi.til1, newpage = FALSE)
plot.new() 
print(dss.epi.til2, newpage = FALSE)
dev.off()
```


```{r}
pdf("./results/Figures/250203_PGEM_DSS_TILgrade_KMplots_78_samples.pdf", width=7, height=6.5)
plot.new() 
print(dss.til, newpage = FALSE)
dev.off()
```


```{r}
pdf("./results/Figures/250203_PGEM_DSS_EpiDISH.RPC_KMplots_78_samples.pdf", width=7, height=6.5)
plot.new() 
print(dss.rpc, newpage = FALSE)
dev.off()
```


## PGEM DSS Barplot using Epi-TIL

```{r, fig.width=15, fig.height=3}
group.vars <- c("Pred", "EpiDISH.RPC.TotalT", "til_grade", "log_Pred") # variable of interest
group.vars2 <- c("Epi-TIL pred", "EpiDISH.RPC", "TIL grade", "Epi-TIL pred (log10)")
cox.vars <- c("age", "sex", "stage")

all.plots <- list()

for (i in 1:length(group.vars)) {
  tmp <- predictions[, c(cox.vars, "dwd_time", "dwd_event", group.vars[i])]
  colnames(tmp)[ncol(tmp)] <- c("variable")
  
  # model 1: put variable at last in the model
  mod.1 <- anova(coxph(Surv(dwd_time, dwd_event) ~ age + sex + stage + variable, data = tmp), test = "Chisq")
  # model 2: put variable first in the model
  mod.2 <- anova(coxph(Surv(dwd_time, dwd_event) ~ variable + age + sex + stage, data = tmp), test = "Chisq")
  
  clin.alone <- mod.1$loglik[4] - mod.1$loglik[1] #all clinical vars only
  clin.add <- mod.2$loglik[5] - mod.2$loglik[2] #
  variable.add <- mod.1$loglik[5] - mod.1$loglik[4] #variable added
  overlap.clin.variable <- mod.1$loglik[5] - mod.1$loglik[1] - clin.add - variable.add
  
  if (overlap.clin.variable < 0) {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.alone, 0, variable.add),
      stringsAsFactors = F
    )
  } else {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.add, overlap.clin.variable, variable.add),
      stringsAsFactors = F
      )
  }
  
  fill.1 <- c("#252525", "grey", "white")
  names(fill.1) <- dat.1$cat
  dat.1$cat <- factor(dat.1$cat, levels = rev(dat.1$cat))
  dat.1 <- melt(dat.1, id.vars = "cat")
  dat.1$variable <- plyr::mapvalues(dat.1$variable, 
                                    from = c("stageOnly", "plusVariable"),
                                    to = c("Clinical variable", 
                                           paste("Clinical variable and ", group.vars[i], sep = "")))
  dat.1$test <- group.vars[i]
  
  # Make ggplot bar plots
  plot.1 <- ggplot(dat.1, aes(x = variable, y = value, fill = cat)) + 
    geom_bar(position = "stack", stat = "identity", color = "black") +
    scale_fill_manual("", values = fill.1, 
                      breaks = names(fill.1), 
                      labels = c("Clinical variables", 
                                 "Overlap", 
                                 group.vars2[i])) +
    scale_y_continuous(limits = c(0, 30)) +
    ggtitle(paste0(group.vars2[i], ": P value ", signif(mod.1$`Pr(>|Chi|)`[5], digits = 3))) +
    theme_45() +
    labs(x = "", fill = "Category", y = "Delta logLik") +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank()#,
          #legend.position = "bottom"
          ) #+
    #guides(fill=guide_legend(nrow=3,byrow=TRUE))
  all.plots[[i]] <- plot.1
  
}
```

## Save ANOVA bar plots

```{r, fig.width=5, fig.height=3.5}
pdf("./results/Figures/250203_PGEM_DSS_EpiTIL_ANOVA_78_samples.pdf", width=5, height=3.5)
all.plots[[1]] 
dev.off()
```

```{r, fig.width=5, fig.height=3.5}
pdf("./results/Figures/250203_PGEM_DSS_EpiDISH.RPC_ANOVA_78_samples.pdf", width=5, height=3.5)
all.plots[[2]]
dev.off()
```

```{r, fig.width=5, fig.height=3.5}
pdf("./results/Figures/250203_PGEM_DSS_TILgrade_ANOVA_78_samples.pdf", width=5, height=3.5)
all.plots[[3]] 
dev.off()
```

```{r, fig.width=5, fig.height=3.5}
pdf("./results/Figures/250203_PGEM_DSS_EpiTIL_log10_ANOVA_78_samples.pdf", width=5, height=3.5)
all.plots[[4]] 
dev.off()
```

# PGEM: Epi-TIL predictors to PFS

```{r}
```


## PFS KM by Epi-TIL tertiles

```{r, calc-PFS-pval}
tmp.logrank <- survival::survdiff(Surv(prog_time, (prog_event | dwd_event )) ~ Pred.tertile2, data=predictions)
tmp.p.logrank <- 1 - pchisq(tmp.logrank$chisq, length(tmp.logrank$n) - 1)
tmp.p.logrank
```


```{r, fig.width=7, fig.height=6.5}
pred.surv <- survival::survfit(Surv(prog_time, (prog_event | dwd_event )) ~ Pred.tertile2, data=predictions)
names(pred.surv$strata) <- gsub("Pred.tertile2=", "", names(pred.surv$strata))
pfs.epi.til1 <- ggsurvplot(fit=pred.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #palette = "Dark2",
                                palette = c("chartreuse4", "darkblue", "firebrick"),
                                ylab="Probability of PFS",
                                xlab="Time (months)", 
                                legend.title = "Epi-TIL predictions")
plot.new() 
print(pfs.epi.til1, newpage = FALSE)
```


```{r, fig.width=7, fig.height=6.5}
pred.surv <- survival::survfit(Surv(prog_time, (prog_event | dwd_event )) ~ Pred.tertile2.log, data=predictions)
names(pred.surv$strata) <- gsub("Pred.tertile2.log=", "", names(pred.surv$strata))
pfs.epi.til2 <- ggsurvplot(fit=pred.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #palette = "Dark2",
                                palette = c("chartreuse4", "darkblue", "firebrick"),
                                ylab="Probability of PFS",
                                xlab="Time (months)", 
                                legend.title = "Epi-TIL (log10) predictions")
plot.new() 
print(pfs.epi.til2, newpage = FALSE)
```



## Save KM plots

```{r}
pdf("./results/Figures/250203_PGEM_PFS_EpiTIL_KMplots_78_samples.pdf", width=7, height=6.5)
plot.new() 
print(pfs.epi.til1, newpage = FALSE)
plot.new() 
print(pfs.epi.til2, newpage = FALSE)
dev.off()
```


## PGEM PFS Barplot using Epi-TIL

```{r}
#group.vars <- c("Pred", "EpiDISH.RPC.TotalT", "til_grade", "log_Pred") # variable of interest
group.vars2 <- c("Epi-TIL pred", "EpiDISH.RPC", "TIL grade", "Epi-TIL pred (log10)")

pfs.plots <- list()
for (i in 1:length(group.vars)) {
  tmp <- predictions[, c(cox.vars, "prog_time", "prog_event", "dwd_event", group.vars[i])]
  colnames(tmp)[ncol(tmp)] <- c("variable")
  
  # model 1: put variable at last in the model
  mod.1 <- anova(coxph(Surv(prog_time, (prog_event | dwd_event )) ~ age + sex + stage + variable, data = tmp), test = "Chisq")
  # model 2: put variable first in the model
  mod.2 <- anova(coxph(Surv(prog_time, (prog_event | dwd_event )) ~ variable + age + sex + stage, data = tmp), test = "Chisq")
  
  clin.alone <- mod.1$loglik[4] - mod.1$loglik[1] #stage only
  clin.add <- mod.2$loglik[5] - mod.2$loglik[2] #
  variable.add <- mod.1$loglik[5] - mod.1$loglik[4] #variable - stage
  overlap.clin.variable <- mod.1$loglik[5] - mod.1$loglik[1] - clin.add - variable.add
  
  if (overlap.clin.variable < 0) {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.alone, 0, variable.add),
      stringsAsFactors = F
    )
  } else {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.add, overlap.clin.variable, variable.add),
      stringsAsFactors = F
      )
  }
  
  fill.1 <- c("#252525", "grey", "white")
  names(fill.1) <- dat.1$cat
  dat.1$cat <- factor(dat.1$cat, levels = rev(dat.1$cat))
  dat.1 <- melt(dat.1, id.vars = "cat")
  dat.1$variable <- plyr::mapvalues(dat.1$variable, 
                                    from = c("stageOnly", "plusVariable"),
                                    to = c("Clinical variable", 
                                           paste("Clinical variable and ", group.vars[i], sep = "")))
  dat.1$test <- group.vars[i]
  
  # Make ggplot bar plots
  plot.1 <- ggplot(dat.1, aes(x = variable, y = value, fill = cat)) + 
    geom_bar(position = "stack", stat = "identity", color = "black") +
    scale_fill_manual("", values = fill.1, 
                      breaks = names(fill.1), 
                      labels = c("Clinical variables", 
                                 "Overlap", 
                                 group.vars2[i])) +
    scale_y_continuous(limits = c(0, 30)) +
    ggtitle(paste0(group.vars2[i], ": P value ", signif(mod.1$`Pr(>|Chi|)`[5], digits = 3))) +
    theme_45() +
    labs(x = "", fill = "Category", y = "Delta logLik") +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
  
  pfs.plots[[i]] <- plot.1
  
}
```

## Save ANOVA bar plots


```{r, fig.width=5, fig.height=3.5}
pdf("./results/Figures/250203_PGEM_PFS_EpiTIL_ANOVA_78_samples.pdf", width=5, height=3.5)
pfs.plots[[1]] 
dev.off()
```




# TCGA: Epi-TIL predictors to DSS

## Set up 

```{r, load-imputed-tcga}
tcga.xb.impute <- readRDS(file="/datastore/nextgenout5/share/labs/bioinformatics/siyao/projects/Thomas-Edmiston_20240711_PredictTILs/results/TCGA_SKCAM_tcga.xb.impute.rds")
#dim(tcga.xb.impute$data)
##[1] 452512    298

dat_imputed <- data.frame(tcga.xb.impute$data, check.names = F)
#dim(dat_imputed)
##[1] 452512    298
#sum(is.na(dat_imputed))
##[1] 0
# tcga.xb[is.na(tcga.xb)] <- 0.5

# clinical
tcga.clinical <- data.table::fread(paste0(params$tcga_dir, "TCGA-CDR-SupplementalTableS1.txt"))

# Convert days to months
tcga.clinical$PFI.time <- tcga.clinical$PFI.time/30
tcga.clinical$OS.time <- tcga.clinical$OS.time/30
tcga.clinical$DSS.time <- tcga.clinical$DSS.time/30

### Only select primary tumor 298 in total
### https://portal.gdc.cancer.gov/ as of 04/19/2022
tcga.primary <- data.table::fread(paste0(params$tcga_dir,"sample.tsv"))
tcga.primary <- tcga.primary %>% filter(sample_type == "Primary Tumor")
tcga.primary <- unique(tcga.primary$case_submitter_id)
#length(tcga.primary)
##[1] 298

# subset clinical data
tcga.clinical <- tcga.clinical[match(colnames(dat_imputed), tcga.clinical$bcr_patient_barcode), ]
#dim(tcga.clinical)
##[1] 298  32
```


```{r, subset}
# Subset data
tmpd <- dat_imputed[ rownames(dat_imputed) %in% rownames(model.probe), ]
#dim(tmpd)

# Subset weights and convert tibble to named vector
tmpp <- model.probe[model.probe$probe %in% rownames(tmpd), ]
#dim(tmpp)
tmpp2 <- as.numeric(tmpp$s1)
names(tmpp2) <- tmpp$probe    
#head(tmpp2)
#head(rownames(tmpd))

# Check if the subsetted probes are in the same order
#identical(names(tmpp2), rownames(tmpd))
## FALSE

# Reorder
tmpp2 <- tmpp2[match(rownames(tmpd), names(tmpp2))]
#identical(names(tmpp2), rownames(tmpd))
##TRUE

# Prediction calculation
datp <- t(tmpd) %*% tmpp2
datp <- datp + as.numeric(model.probe[model.probe$probe == "(Intercept)", "s1"])
#dim(datp)

# Reformat datp
datp_table <- data.frame("Sample_Name"=rownames(datp), 
                         "glmnet.pred" = as.numeric(datp))
row.names(datp_table) <- NULL
#head(datp_table)

#hist(datp_table$glmnet.pred)
#summary(datp_table$glmnet.pred)

datp_table$glmnet.pred.cat <- cut(datp_table$glmnet.pred, 
                                  c(-9999, quantile(datp_table$glmnet.pred, c(0.33, 0.66)),
                                    99999), 
                                  labels = c("low", "med", "high"))

datp_table$glmnet.pred.tertile = ntile(datp_table$glmnet.pred, 3)
#table(datp_table$glmnet.pred.tertile)
##1   2   3 
##141 141 140
datp_table$glmnet.pred.tertile2 <- recode(datp_table$glmnet.pred.tertile, 
                                       "1"="low", "2"="med", "3"="high" )
#table(datp_table$glmnet.pred.cat, datp_table$glmnet.pred.tertile2)
```


```{r, prep-predictions}
predictions <- data.frame(
  Sample = tcga.clinical$bcr_patient_barcode,
  prog_event = tcga.clinical$PFI,
  prog_time = tcga.clinical$PFI.time,
  os_event = tcga.clinical$OS,
  os_time = tcga.clinical$OS.time,
  dss_event = tcga.clinical$DSS,
  dss_time = tcga.clinical$DSS.time,
  age = tcga.clinical$age_at_initial_pathologic_diagnosis,
  stage = tcga.clinical$ajcc_pathologic_tumor_stage,
  sex = tcga.clinical$gender,
  stringsAsFactors = F
)
predictions$stage[predictions$stage %in% c("[Not Available]", "I/II NOS")] <- NA
all(predictions$Sample == datp_table$Sample_Name)
##TRUE
```

```{r, combine-meta}
predictions <- cbind(predictions, datp_table)
predictions$sex <- factor(predictions$sex, levels = c("MALE", "FEMALE"))
predictions$stage[predictions$stage == "Stage 0"] <- NA

### have to use combined stage as there are unclear stages (22 stage III that don't know if IIIA or IIIB; 17 Stage II that don't know if IIA or IIB)
predictions$stage <- plyr::mapvalues(predictions$stage,
                                     from = c("Stage IA", "Stage IB", "Stage IIA", "Stage IIB", "Stage IIC", "Stage IIIA", "Stage IIIB", "Stage IIIC"),
                                     to = c("Stage I", "Stage I", "Stage II", "Stage II", "Stage II", "Stage III", "Stage III", "Stage III"))
predictions$stage <- factor(predictions$stage, 
                            levels = c("Stage I", "Stage II", "Stage III", "Stage IV"), ordered = T)

predictions$glmnet.pred.tertile2 <- factor(predictions$glmnet.pred.tertile2,
                                           levels=c("low","med","high"))
sum(is.na(predictions$os_time))
##[9]
sum(is.na(predictions$dss_time))
##[9]
sum(is.na(predictions$prog_time))
##[8]
```


```{r, load-epidish}
m.decon <- readRDS(file="./results/TCGA_SKCM_EpiDISHOutput.rds")
predictions$EpiDISH.RPC.TotalT <- m.decon[[1]]$TotalT
predictions$EpiDISH.CBS.TotalT <- m.decon[[2]]$TotalT
predictions$EpiDISH.CP.TotalT <- m.decon[[3]]$TotalT

predictions$EpiDISH.RPC.TotalT.tertile <- m.decon[[1]]$TotalT.tertile2
predictions$EpiDISH.CBS.TotalT.tertile <- m.decon[[2]]$TotalT.tertile2
predictions$EpiDISH.CP.TotalT.tertile <- m.decon[[3]]$TotalT.tertile2
```


## Run survival analysis

```{r, plotDSS, fig.width=7, fig.height=6.5}
pred.surv <- survival::survfit(Surv(dss_time, dss_event) ~ glmnet.pred.tertile2, data=predictions)
names(pred.surv$strata) <- gsub("glmnet.pred.tertile2=", "", names(pred.surv$strata))
pred.p <- ggsurvplot(fit=pred.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #pval = signif(tmp.p.logrank,3), 
                                palette = c("chartreuse4", "darkblue", "firebrick"), 
                                #palette = "Dark2",
                                ylab="Probability of DSS",
                                xlab="Time (months)", 
                                #xlim = c(0, 60), break.time.by = 12,
                                legend.title = "Predicted TIL")
pred.p
```


## Save KM plots

```{r}
pdf("./results/Figures/250203_TCGA_DSS_EpiTIL_KMplots.pdf", width=7, height=6.5)
plot.new() 
print(pred.p, newpage = FALSE)
dev.off()
```

## ANOVA LRT

```{r, run-dss-LRT-anova}
group.vars <- c("glmnet.pred.tertile2") # variable of interest
group.vars2 <- c("Epi-TIL pred")
cox.vars <- c("age", "sex", "stage")

all.plots <- list()

for (i in 1:length(group.vars)) {
  tmp <- predictions[, c(cox.vars, "dss_time", "dss_event", group.vars[i])]
  colnames(tmp)[ncol(tmp)] <- c("variable")
  
  # model 1: put variable at last in the model
  mod.1 <- anova(coxph(Surv(dss_time, dss_event) ~ age + sex + stage + variable, data = tmp), test = "Chisq")
  # model 2: put variable first in the model
  mod.2 <- anova(coxph(Surv(dss_time, dss_event) ~ variable + age + sex + stage, data = tmp), test = "Chisq")
  
  clin.alone <- mod.1$loglik[4] - mod.1$loglik[1] #all clinical vars only
  clin.add <- mod.2$loglik[5] - mod.2$loglik[2] #
  variable.add <- mod.1$loglik[5] - mod.1$loglik[4] #variable added
  overlap.clin.variable <- mod.1$loglik[5] - mod.1$loglik[1] - clin.add - variable.add
  
  if (overlap.clin.variable < 0) {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.alone, 0, variable.add),
      stringsAsFactors = F
    )
  } else {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.add, overlap.clin.variable, variable.add),
      stringsAsFactors = F
      )
  }
  
  fill.1 <- c("#252525", "grey", "white")
  names(fill.1) <- dat.1$cat
  dat.1$cat <- factor(dat.1$cat, levels = rev(dat.1$cat))
  dat.1 <- melt(dat.1, id.vars = "cat")
  dat.1$variable <- plyr::mapvalues(dat.1$variable, 
                                    from = c("stageOnly", "plusVariable"),
                                    to = c("Clinical variable", 
                                           paste("Clinical variable and ", group.vars[i], sep = "")))
  dat.1$test <- group.vars[i]
  
  # Make ggplot bar plots
  plot.1 <- ggplot(dat.1, aes(x = variable, y = value, fill = cat)) + 
    geom_bar(position = "stack", stat = "identity", color = "black") +
    scale_fill_manual("", values = fill.1, 
                      breaks = names(fill.1), 
                      labels = c("Clinical variables", 
                                 "Overlap", 
                                 group.vars2[i])) +
    scale_y_continuous(limits = c(0, 30)) +
    ggtitle(paste0(group.vars2[i], ": P value ", signif(mod.1$`Pr(>|Chi|)`[5], digits = 3))) +
    theme_45() +
    labs(x = "", fill = "Category", y = "Delta logLik") +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
  
  all.plots[[i]] <- plot.1
  
}
```

## Save ANOVA bar plots

```{r}
pdf("./results/Figures/250203_TCGA_DSS_EpiTIL_ANOVA.pdf", width=5, height=3.5)
all.plots[[1]] 
dev.off()
```


# TCGA: Epi-TIL predictors to PFS

## Run survival analysis

```{r, plotPFS, fig.width=7, fig.height=6.5}
pred.surv <- survival::survfit(Surv(prog_time, prog_event) ~ glmnet.pred.tertile2, data=predictions)
names(pred.surv$strata) <- gsub("glmnet.pred.tertile2=", "", names(pred.surv$strata))
pred.pfs <- ggsurvplot(fit=pred.surv, 
                                data=predictions,
                                risk.table = TRUE, 
                                pval = TRUE,
                                #palette = "Dark2",
                                palette = c("chartreuse4", "darkblue", "firebrick"), 
                                ylab="Probability of PFS",
                                xlab="Time (months)", 
                                legend.title = "Predicted TIL")
pred.pfs
```

## Save KM plots

```{r}
pdf("./results/Figures/250203_TCGA_PFS_EpiTIL_KMplots.pdf", width=7, height=6.5)
plot.new() 
print(pred.pfs, newpage = FALSE)
dev.off()
```

## ANOVA LRT

```{r, run-pfs-LRT-anova}
#group.vars <- c("glmnet.pred.tertile2", "EpiDISH.RPC.TotalT") # variable of interest
#group.vars2 <- c("Epi-TIL pred", "EpiDISH.RPC.TotalT")
#cox.vars <- c("age", "sex", "stage")
group.vars <- c("glmnet.pred.tertile2") # variable of interest
group.vars2 <- c("Epi-TIL pred")
cox.vars <- c("age", "sex", "stage")
all.plots <- list()

for (i in 1:length(group.vars)) {
  tmp <- predictions[, c(cox.vars, "prog_time", "prog_event", group.vars[i])]
  colnames(tmp)[ncol(tmp)] <- c("variable")
  
  # model 1: put variable at last in the model
  mod.1 <- anova(coxph(Surv(prog_time, prog_event) ~ age + sex + stage + variable, data = tmp), test = "Chisq")
  # model 2: put variable first in the model
  mod.2 <- anova(coxph(Surv(prog_time, prog_event) ~ variable + age + sex + stage, data = tmp), test = "Chisq")
  
  clin.alone <- mod.1$loglik[4] - mod.1$loglik[1] #all clinical vars only
  clin.add <- mod.2$loglik[5] - mod.2$loglik[2] #
  variable.add <- mod.1$loglik[5] - mod.1$loglik[4] #variable added
  overlap.clin.variable <- mod.1$loglik[5] - mod.1$loglik[1] - clin.add - variable.add
  
  if (overlap.clin.variable < 0) {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.alone, 0, variable.add),
      stringsAsFactors = F
    )
  } else {
    dat.1 <- data.frame(
      cat = c("clinicalVar", "overlap", "variable"),
      stageOnly = c(clin.alone, 0, 0),
      plusVariable = c(clin.add, overlap.clin.variable, variable.add),
      stringsAsFactors = F
      )
  }
  
  fill.1 <- c("#252525", "grey", "white")
  names(fill.1) <- dat.1$cat
  dat.1$cat <- factor(dat.1$cat, levels = rev(dat.1$cat))
  dat.1 <- melt(dat.1, id.vars = "cat")
  dat.1$variable <- plyr::mapvalues(dat.1$variable, 
                                    from = c("stageOnly", "plusVariable"),
                                    to = c("Clinical variable", 
                                           paste("Clinical variable and ", group.vars[i], sep = "")))
  dat.1$test <- group.vars[i]
  
  # Make ggplot bar plots
  plot.1 <- ggplot(dat.1, aes(x = variable, y = value, fill = cat)) + 
    geom_bar(position = "stack", stat = "identity", color = "black") +
    scale_fill_manual("", values = fill.1, 
                      breaks = names(fill.1), 
                      labels = c("Clinical variables", 
                                 "Overlap", 
                                 group.vars2[i])) +
    scale_y_continuous(limits = c(0, 30)) +
    ggtitle(paste0(group.vars2[i], ": P value ", signif(mod.1$`Pr(>|Chi|)`[5], digits = 3))) +
    theme_45() +
    labs(x = "", fill = "Category", y = "Delta logLik") +
    theme(axis.text.x=element_blank(), axis.ticks.x=element_blank())
  
  all.plots[[i]] <- plot.1
  
}
```

## Save ANOVA bar plots

```{r}
pdf("./results/Figures/250203_TCGA_PFS_EpiTIL_ANOVA.pdf", width=5, height=3.5)
all.plots[[1]] 
dev.off()
```



# InterMEL

## Load data

```{r}
load(params$intermel.data)
load(params$intermel.clin)
iml_clin_dat$tils_dictionary_key <- factor(iml_clin_dat$tils_dictionary_key, levels = c("Brisk", "Non-brisk", "Absent", "Cannot assess"))

final.probe <- read.table(file=params$final_probes, sep="\t", header=T)
dim(final.probe)
##[1] 169   1
final.probe$probe <- rownames(final.probe)

table(rownames(dat_imputed) %in% final.probe$probe)
##FALSE   TRUE 
##432511     81

identical(colnames(dat_imputed), rownames(sams))
##TRUE

sub.clin <- iml_clin_dat[iml_clin_dat$Sample_Name %in% sams$Sample_Name, ]
dim(sub.clin)
##[1] 422  95

table(sub.clin$vital_status_dictionary_key)
##Alive       Dead of melanoma   Dead of other causes 
##  188                    203                     22 
##Dead of unknown causes
##  9
table(sub.clin$group_dictionary_key)

# Match sub.clin sample order with sams
identical(sub.clin$Sample_Name, sams$Sample_Name) ##FALSE
sub.clin <- sub.clin[match(sams$Sample_Name, sub.clin$Sample_Name), ]
identical(sub.clin$Sample_Name, sams$Sample_Name)

tmpd <- dat_imputed[rownames(dat_imputed) %in% rownames(final.probe), ]
# Subset weights and convert tibble to named vector
tmpp <- final.probe[final.probe$probe %in% rownames(tmpd), ]
tmpp2 <- as.numeric(tmpp$s1)
names(tmpp2) <- tmpp$probe    
identical(names(tmpp2), rownames(tmpd))
##TRUE

# Prediction calculation
datp <- tmpp2 %*% tmpd
datp <- datp + as.numeric(final.probe[final.probe$probe == "(Intercept)", "s1"])
# Reformat datp
datp_table <- data.frame("Sample_Name"=colnames(datp), 
                         "glmnet.pred.log" = as.numeric(datp),
                         "glmnet.pred"=10^(as.numeric(datp) - 0.01))
row.names(datp_table) <- colnames(datp)

# check if samples are in the same order
identical(datp_table$Sample_Name, rownames(sams))
```

## Epi-TIL boxplot

```{r, fig.width=6, fig.height=5}
plot.df <- cbind(datp_table, sub.clin[, c( "tils_dictionary_key", "group_dictionary_key")])
plot.df$group_dictionary_key <- recode(plot.df$group_dictionary_key, "Died from melanoma within 5 years of diagnosis"="Die within 5 yrs",
                                       "Lived more than 5 years after diagnosis"="Survive longer 5 yrs")
dim(plot.df)
##[1] 422 4
sum(is.na(plot.df$tils_dictionary_key))
##[21]
table(plot.df$tils_dictionary_key)

epitil.p1 <- ggplot(plot.df, 
       aes(x=group_dictionary_key, y=glmnet.pred, fill=group_dictionary_key)) +
  geom_boxplot() + 
  geom_point(size = 1, color="grey35", position = position_jitterdodge()) +
  ggtitle("Epi-TIL final probes: \n81 out of 168 present in InterMEL") +
  scale_fill_manual("", values=c("red", "white")) +
  theme_0() +
  theme(legend.position = "none") +
  xlab("") + 
  ylab("Predicted Epi-TIL") +
  stat_compare_means()
epitil.p1 + labs(tag="A")

epitil.p2 <- ggplot(plot.df, 
       aes(x=group_dictionary_key, y=glmnet.pred.log, fill=group_dictionary_key)) +
  geom_boxplot() + 
  geom_point(size = 1, color="grey35", position = position_jitterdodge()) +
  ggtitle("Epi-TIL final probes: \n81 out of 168 present in InterMEL") +
  scale_fill_manual("", values=c("red", "white")) +
  theme_0() +
  theme(legend.position = "none") +
  xlab("") + 
  ylab("Predicted Epi-TIL (log10)") +
  stat_compare_means()
epitil.p2 + labs(tag="A")
```


## EpiDISH.RPC boxplot

```{r, fig.width=6, fig.height=5}
m.decon <- readRDS(file="./results/InterMEL_EpiDISHOutput.rds")
tmp <- m.decon$EpiDISH_RPC
identical(rownames(tmp), rownames(sams))
plot.df <- cbind(tmp, 
                 sub.clin[, c( "tils_dictionary_key", "group_dictionary_key")])

plot.df$group_dictionary_key <- recode(plot.df$group_dictionary_key, "Died from melanoma within 5 years of diagnosis"="Die within 5 yrs",
                                       "Lived more than 5 years after diagnosis"="Survive longer 5 yrs")
epidish.p <- ggplot(plot.df, 
       aes(x=group_dictionary_key, y=TotalT, fill=group_dictionary_key)) +
  geom_boxplot() + 
  geom_point(size = 1, color="grey35", position = position_jitterdodge()) +
  ggtitle(names(m.decon)[1]) +
  scale_fill_manual("", values=c("red", "white")) +
  theme_0() +
  theme(legend.position = "none") +
  xlab("") + 
  ylab("Total T") +
  stat_compare_means()
epidish.p + labs(tag="B")
```

## Barplot of Epi-TIL tertiles by alive and dead

```{r, barplot-by-survival, fig.width=6.5, fig.height=4}
plot.df <- cbind(datp_table, sub.clin[, c( "tils_dictionary_key", "group_dictionary_key")])
plot.df$group_dictionary_key <- recode(plot.df$group_dictionary_key, "Died from melanoma within 5 years of diagnosis"="Die within 5 yrs",
                                       "Lived more than 5 years after diagnosis"="Survive longer 5 yrs")


# divide glmnet.pred into high low and medium
plot.df$glmnet.pred.tertile <- dplyr::ntile(plot.df$glmnet.pred, 3)
table(plot.df$glmnet.pred.tertile)
##1   2   3 
##141 141 140
plot.df$glmnet.pred.tertile2 <- recode(plot.df$glmnet.pred.tertile, 
                                       "1"="low", "2"="interm", "3"="high" )

plot.df$glmnet.pred.tertile.log <- dplyr::ntile(plot.df$glmnet.pred.log, 3)
table(plot.df$glmnet.pred.tertile.log)
##1   2   3 
##141 141 140
plot.df$glmnet.pred.tertile2.log <- recode(plot.df$glmnet.pred.tertile.log, 
                                       "1"="low", "2"="interm", "3"="high" )

plot.df2 <- as.data.frame(table(plot.df$group_dictionary_key, plot.df$glmnet.pred.tertile2))
epitil.bar1 <- ggplot(plot.df2, aes(Var1, Freq, fill=Var2)) +
  geom_bar(position="fill", stat="identity", color="white") + 
  scale_fill_manual("Epi-TIL tertiles", 
                    #values=c("high"="#3B9AB2", "interm"="#EBCC2A", "low"="#F21A00"),
                    values=c("low"="chartreuse4", "interm"="darkblue", "high"="firebrick")) +
  #scale_fill_tron() +
  theme_0() + 
  xlab("") +
  ylab("Proportion of samples") + 
  ggtitle("Epi-TIL final 81 probes")
epitil.bar1 

plot.df2 <- as.data.frame(table(plot.df$group_dictionary_key, plot.df$glmnet.pred.tertile2.log))
epitil.bar2 <- ggplot(plot.df2, aes(Var1, Freq, fill=Var2)) +
  geom_bar(position="fill", stat="identity", color="white") + 
  scale_fill_manual("Epi-TIL (log10) tertiles", 
                    #values=c("high"="#3B9AB2", "interm"="#EBCC2A", "low"="#F21A00"),
                    values=c("low"="chartreuse4", "interm"="darkblue", "high"="firebrick")) +
  #scale_fill_tron() +
  theme_0() + 
  xlab("") +
  ylab("Proportion of samples") + 
  ggtitle("Epi-TIL final 81 probes")
epitil.bar2 
```

## Barplot of TIL grade by alive and dead

```{r, fig.width=6.5, fig.height=4}
table(plot.df$tils_dictionary_key)
##           Brisk     Non-brisk       Absent      Cannot assess 
##           18           336            45             2
plot.df3 <- as.data.frame(table(plot.df$group_dictionary_key, plot.df$tils_dictionary_key))
plot.df3 <- plot.df3[plot.df3$Var2!="Cannot assess",]
plot.df3$Var2 <- factor(plot.df3$Var2, levels=c( "Brisk", "Non-brisk","Absent"))

til.bar <- ggplot(plot.df3, aes(Var1, Freq, fill=Var2)) +
  geom_bar(position="fill", stat="identity", color="white") + 
  scale_fill_manual("TIL grade", 
                    #values=c("Brisk"="#3B9AB2", "Non-brisk"="#EBCC2A", "Absent"="#F21A00"),
                    values=c("Absent"="chartreuse4", "Non-brisk"="darkblue", "Brisk"="firebrick")) +
  theme_0() + 
  xlab("") +
  ylab("Proportion of samples") + 
  ggtitle("TIL grade")
til.bar + labs(tag="D")
```



## Save plots

```{r, fig.width=6, fig.height=5}
pdf("./results/Figures/250203_InterMEL_EpiTIL_Boxplot.pdf", width=6, height=5)
epitil.p1
epitil.p2
dev.off()
```


```{r, fig.width=6, fig.height=5}
pdf("./results/Figures/250203_InterMEL_EpiDISH_Boxplot.pdf", width=6, height=5)
epidish.p
dev.off()
```



```{r, fig.width=6, fig.height=4.5}
pdf("./results/Figures/250203_InterMEL_TILgrade_Barplot.pdf", width=6, height=4.5)
til.bar
dev.off()
```


```{r, fig.width=6.5, fig.height=4}
pdf("./results/Figures/250203_InterMEL_EpiTIL_Barplot.pdf", width=6.5, height=4)
epitil.bar1
epitil.bar2
dev.off()
```


## Epi-TIL scores by case/control in each TIL grade

```{r, fig.width=8, fig.height=6}
plot.df4 <- plot.df[!is.na(plot.df$tils_dictionary_key),]
p <- ggplot(plot.df4[plot.df4$tils_dictionary_key!="Cannot assess", ], 
       aes(x=tils_dictionary_key, y=glmnet.pred, fill=group_dictionary_key)) +
  geom_boxplot(outlier.size=0.01, outlier.shape=NA) + 
  #geom_jitter(width=0.1, color="grey35") +
  geom_point(position = position_jitterdodge(jitter.width=0.1), alpha=0.3) +
  ggtitle("Epi-TIL final probes: \n81 out of 168 present in InterMEL") +
  scale_fill_manual("", values=c("red", "white")) +
  theme_0() +
  xlab("") + 
  ylab("Predicted Epi-TIL") +
  ggpubr::stat_compare_means(aes(group = group_dictionary_key), label = "p.format")
p
```

```{r}
pdf("./results/Figures/250226_InterMEL_EpiTIL_bySurvivalbyTILgrade.pdf", width=8, height=6)
p
dev.off()
```


# GO terms for the 81 probes/CpGs


## Map selected probes to genes. Are they associated with infiltrating T cells?

```{r}
#if (!require("BiocManager", quietly = TRUE))
#  install.packages("BiocManager")
#BiocManager::install("IlluminaHumanMethylation450kanno.ilmn12.hg19", force=TRUE)
#library(IlluminaHumanMethylation450kanno.ilmn12.hg19)
library(IlluminaHumanMethylationEPICv2anno.20a1.hg38)
```


## Get annotations for all probes in the data

```{r}
#data("IlluminaHumanMethylation450kanno.ilmn12.hg19")
#data("IlluminaHumanMethylationEPICanno.ilm10b2.hg19")
data("IlluminaHumanMethylationEPICv2anno.20a1.hg38")
#anno <- getAnnotation(IlluminaHumanMethylationEPICanno.ilm10b2.hg19) # 
anno <- getAnnotation(IlluminaHumanMethylationEPICv2anno.20a1.hg38)
anno.names <- substr(rownames(anno), 1, 10)
head(anno.names)
anno$anno.names <- anno.names
dim(anno)
##[1] 930075     43
anno <- anno[ anno.names %in% rownames(dat), ]# intermel 850k probes
dim(anno)
##[1] 723341     43
```


```{r}
table(rownames(dat_imputed) %in% final.probe$probe)
##FALSE   TRUE 
##432511     81

model.probe <- final.probe[final.probe$probe%in% rownames(dat_imputed), ]
dim(model.probe)
head(rownames(model.probe)[-1])

final.model <- merge(model.probe, anno, by.x = "probe", by.y="anno.names")
dim(final.model)
##79 --> 2 probes do not have annotation
final.model$Gene <- unlist(lapply(final.model$UCSC_RefGene_Name, function(x){
  x <- strsplit(x, split = ";")[[1]]
  x <- unique(x)
  if(length(x) == 0){
    return(NA)
  }else{
    return(paste(x, collapse = ",", sep = ""))
  }
}))
dim(final.model)
##[1] 79 45
```


### GO term enrichment analysis 

```{r}
all.genes <- anno[anno$UCSC_RefGene_Name != "", ]
all.genes <- lapply(all.genes$UCSC_RefGene_Name, function(x){
  x <- strsplit(x, split = ";")[[1]]
  x <- unique(x)
  return(x)
})
all.genes <- unique(unlist(all.genes))
length(all.genes)
##[1]  19817
head(all.genes)
##[1] "MAEL"     "NR5A2"    "C1orf114" "RFX5"     "TNNT2"    "ZNF362"
degs <- final.model
degs <- lapply(degs$UCSC_RefGene_Name, function(x){
  x <- strsplit(x, split = ";")[[1]]
  x <- unique(x)
  return(x)
})
degs <- unique(unlist(degs))
length(degs)
##63
head(degs)
##[1] "UXS1"   "SHISA9" "VGLL4"  "EIF2B1" "GTF2H3" "SQSTM1"
```


### GO terms

```{r}
library(topGO)
library(org.Hs.eg.db)
library(grid)
library(gridExtra)

geneList <- ifelse(all.genes %in% degs, 1, 0)
names(geneList) <- all.genes

# Create topGOdata object
GOdata <- new("topGOdata",
              ontology = "BP", # use biological process ontology
              allGenes = geneList,
              geneSelectionFun = function(x)(x == 1),
              annot = annFUN.org, mapping = "org.Hs.eg.db", ID = "symbol")

# Run Fisher's test
resultFisher <- runTest(GOdata, algorithm = "elim", statistic = "fisher")
# Extract table
go.res <- GenTable(GOdata, Fisher = resultFisher, 
                   topNodes = length(resultFisher@score), numChar = 60)
```


### Plot Top 20 GO terms associated with the 81 probes in the final Epi-TIL model

```{r, fig.width=8, fig.height=5.5}
sub.go <- go.res[go.res$Fisher < 0.01, ]
sub.go$logp <- -log10(as.numeric(sub.go$Fisher))
sub.go <- sub.go[!duplicated(sub.go$Term), ]
sub.go$Term <- factor(sub.go$Term, levels = rev(sub.go$Term))
dim(sub.go)
sub.go$newTerm <- stringr::str_wrap(sub.go$Term, width=80)
p <- ggplot(sub.go, aes(x = newTerm, y = logp)) +
  geom_bar(stat = "identity") +
  scale_y_continuous() +
  labs(x = "", y = "-log10(P)") +
  theme_0() + 
  geom_hline(yintercept=-log10(0.01), lty=2) + 
  coord_flip() + ggtitle("Significant GO terms")
p
```


## Save plots

```{r, fig.width=6, fig.height=5}
pdf("./results/Figures/250203_InterMEL_81_EpiTIL_probes_Top20GoTerms.pdf", width=8, height=5.5)
p
dev.off()
```



# R SESSION INFORMATION

<details>
<summary>Information about R, the OS and attached or loaded packages</summary>

```{r sesion_info}
pander::pander(sessionInfo(), compact = FALSE)
```
</details>
***
<center>`r format(Sys.time(), '%d %B %Y')`</center>

